indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamSamplerSpec.java

```diff
@@ -0,0 +1,202 @@
+import org.apache.druid.indexing.overlord.sampler.FirehoseSampler;
+import org.apache.druid.indexing.overlord.sampler.SamplerConfig;
+import org.apache.druid.indexing.overlord.sampler.SamplerException;
+import org.apache.druid.indexing.overlord.sampler.SamplerResponse;
+import org.apache.druid.indexing.overlord.sampler.SamplerSpec;
+import org.apache.druid.indexing.seekablestream.common.OrderedPartitionableRecord;
+import org.apache.druid.indexing.seekablestream.common.RecordSupplier;
+import org.apache.druid.indexing.seekablestream.common.StreamPartition;
+import org.apache.druid.indexing.seekablestream.supervisor.SeekableStreamSupervisorIOConfig;
+import org.apache.druid.indexing.seekablestream.supervisor.SeekableStreamSupervisorSpec;
+import org.apache.druid.indexing.seekablestream.supervisor.SeekableStreamSupervisorTuningConfig;
+import org.apache.druid.java.util.common.parsers.ParseException;
+import org.apache.druid.segment.indexing.DataSchema;
+import org.apache.druid.utils.Runnables;
+
+import javax.annotation.Nullable;
+import java.io.File;
+import java.nio.ByteBuffer;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public abstract class SeekableStreamSamplerSpec<PartitionIdType, SequenceOffsetType> implements SamplerSpec
+{
+  private static final int POLL_TIMEOUT_MS = 100;
+
+  private final DataSchema dataSchema;
+  private final FirehoseSampler firehoseSampler;
+
+  protected final SeekableStreamSupervisorIOConfig ioConfig;
+  protected final SeekableStreamSupervisorTuningConfig tuningConfig;
+  protected final SamplerConfig samplerConfig;
+
+  public SeekableStreamSamplerSpec(
+      final SeekableStreamSupervisorSpec ingestionSpec,
+      final SamplerConfig samplerConfig,
+      final FirehoseSampler firehoseSampler
+  )
+  {
+    this.dataSchema = Preconditions.checkNotNull(ingestionSpec, "[spec] is required").getDataSchema();
+    this.ioConfig = Preconditions.checkNotNull(ingestionSpec.getIoConfig(), "[spec.ioConfig] is required");
+    this.tuningConfig = ingestionSpec.getTuningConfig();
+    this.samplerConfig = samplerConfig;
+    this.firehoseSampler = firehoseSampler;
+  }
+
+  @Override
+  public SamplerResponse sample()
+  {
+    return firehoseSampler.sample(new FirehoseFactory()
+    {
+      @Override
+      public Firehose connect(InputRowParser parser, @Nullable File temporaryDirectory)
+      {
+        return getFirehose(parser);
+      }
+    }, dataSchema, samplerConfig);
+  }
+
+  protected abstract Firehose getFirehose(InputRowParser parser);
+
+  protected abstract class SeekableStreamSamplerFirehose implements Firehose
+  {
+    private final InputRowParser parser;
+    private final RecordSupplier<PartitionIdType, SequenceOffsetType> recordSupplier;
+
+    private Iterator<OrderedPartitionableRecord<PartitionIdType, SequenceOffsetType>> recordIterator;
+    private Iterator<byte[]> interRecordIterator;
```

Comment:
the starting bit of `nextRowWithRaw` might be a bit clearer if this variable was named something like `recordBytesIterator` or `recordDataIterator`?